R/pit.R

```diff
@@ -0,0 +1,146 @@
+  x <- as_draws_matrix(x)
+  if (!is.null(loo_weights)) {
+    loo_weights <- as_draws_matrix(loo_weights)
+  }
+  pit(x, y, loo_weights, log)
+}
+
+#' @rdname pit
+#' @export
+pit.draws_matrix <- function(x, y, loo_weights = NULL, log = TRUE) {
+  if (length(y) != ncol(x)) {
+    stop_no_call("Length of `y` must match number of columns in `x`")
+  }
+  if (!is.null(loo_weights)) {
+    loo_weights <- validate_loo_weights(loo_weights, x, log)
+    loo_weights <- normalize_log_weights(loo_weights)
+  }
+  pit <- vapply(seq_len(ncol(x)), function(j) {
+    sel_min <- x[, j] < y[j]
+    if (!any(sel_min)) {
+      pit <- 0
+    } else {
+      if (is.null(loo_weights)) {
+        pit <- mean(sel_min)
+      } else {
+        pit <- exp(log_sum_exp(loo_weights[sel_min, j]))
+      }
+    }
+
+    sel_sup <- x[, j] == y[j]
+    if (any(sel_sup)) {
+      # randomized PIT for discrete y (see, e.g., Czado, C., Gneiting, T.,
+      # Held, L.: Predictive model assessment for count data.
+      # Biometrics 65(4), 1254–1261 (2009).)
+      if (is.null(loo_weights)) {
+        pit_sup <- pit + mean(sel_sup)
+      } else {
+        pit_sup <- pit + exp(log_sum_exp(loo_weights[sel_sup, j]))
+      }
+
+      pit <- runif(1, pit, pit_sup)
+    }
+    pit
+  }, FUN.VALUE = 1.0)
+
+  if (any(pit > 1)) {
+    warning_no_call(
+      paste(
+        "Some PIT values larger than 1! ",
+        "This is usually due to numerical inaccuracies. ",
+        "Largest value: ",
+        max(pit),
+        "\nRounding PIT > 1 to 1.",
+        sep = ""
+      )
+    )
+    pit <- pmin(1, pit)
+  }
+
+  pit
+}
+
+#' @rdname pit
+#' @export
+pit.rvar <- function(x, y, loo_weights = NULL, log = TRUE) {
+  if (any(dim(y) != dim(x))) {
+    stop_no_call("Shape of `y` must match that of `x`.")
+  }
+  if (is.null(loo_weights)) {
+    pit <- array(
+      runif(length(y), Pr(x < y), Pr(x <= y)),
+      dim(x),
+      dimnames(x)
+    )
+  } else {
+    pit <- array(
+      pit(
+        as_draws_matrix(c(x)),
+        y,
+        as_draws_matrix(c(loo_weights)),
+        log
+      ),
+      dim(x),
+      dimnames(x)
+    )
+  }
+
+  pit
+}
+
+# internal ----------------------------------------------------------------
+
+# validate weights and return log weights
+validate_loo_weights <- function(loo_weights, draws, log = TRUE) {
+  checkmate::assert_numeric(loo_weights, any.missing = FALSE)
+  checkmate::assert_flag(log)
+  if (!all(dim(loo_weights) == dim(draws))) {
+    stop_no_call("Dimension of `loo_weights` must match that of `x`.")
+  }
+  if (!all(is.finite(loo_weights))) {
+    stop_no_call("All weigths in `loo_weights` must be finite.")
+  }
+  if (!log) {
+    if (any(loo_weights < 0)) {
+      stop_no_call("`loo-weights` must be non-negative when log = FALSE.")
+    }
+    loo_weights <- log(loo_weights)
+  }
+  loo_weights
+}
+
+normalize_log_weights <- function(log_weights) {
+  apply(log_weights, 2, \(col) col - log_sum_exp(col))
```

Comment:
I think the `\(x)` syntax requires R 4.1.0

Reply:
Good catch! Fixed that.